internal/pkg/source/source.go

```diff
@@ -75,104 +75,185 @@ func New(in ssa.Node, config classifier) *Source {
-				continue
-			}
 
-			// If this call's index is lower than the highest in its block,
-			// then this call is "in the past" and we should stop traversing.
-			i, ok := indexInBlock(r)
-			if !ok {
-				continue
-			}
-			if i < maxInstrReached[r.Block()] {
-				continue
-			}
+	switch t := n.(type) {
+	case *ssa.Alloc:
+		// An Alloc represents the allocation of space for a variable. If a Node is an Alloc,
+		// and the thing being allocated is not an array, then either:
+		// a) it is a Source value, in which case it will get its own traversal when sourcesFromBlocks
+		//    finds this Alloc
+		// b) it is not a Source value, in which case we should not visit it.
+		// However, if the Alloc is an array, then that means the source that we are visiting from
+		// is being placed into an array, slice or varags, so we do need to keep visiting.
+		if _, isArray := utils.Dereference(t.Type()).(*types.Array); isArray {
+			s.visitReferrers(n, maxInstrReached, lastBlockVisited)
 		}
 
-		if fa, ok := r.(*ssa.FieldAddr); ok {
-			deref := utils.Dereference(fa.X.Type())
-			typPath, typName := utils.DecomposeType(deref)
-			fieldName := utils.FieldName(fa)
+	case *ssa.Call:
+		if callee := t.Call.StaticCallee(); callee != nil && s.config.IsSanitizer(utils.DecomposeFunction(callee)) {
+			s.sanitizers = append(s.sanitizers, &sanitizer.Sanitizer{Call: t})
+		}
 
-			if !s.config.IsSourceField(typPath, typName, fieldName) {
-				continue
-			}
+		// This is to avoid attaching calls where the source is the receiver, ex:
+		// core.Sinkf("Source id: %v", wrapper.Source.GetID())
+		if recv := t.Call.Signature().Recv(); recv != nil && s.config.IsSourceType(utils.DecomposeType(utils.Dereference(recv.Type()))) {
+			return
+		}
+
+		s.visitReferrers(n, maxInstrReached, lastBlockVisited)
+		s.visitOperands(n, maxInstrReached, lastBlockVisited)
+
+	case *ssa.FieldAddr:
+		deref := utils.Dereference(t.X.Type())
+		typPath, typName := utils.DecomposeType(deref)
+		fieldName := utils.FieldName(t)
+		if !s.config.IsSourceField(typPath, typName, fieldName) {
+			return
 		}
+		s.visitReferrers(n, maxInstrReached, lastBlockVisited)
+		s.visitOperands(n, maxInstrReached, lastBlockVisited)
+
+	// Everything but the actual integer Index should be visited.
+	case *ssa.Index:
+		s.visitReferrers(n, maxInstrReached, lastBlockVisited)
+		s.dfs(t.X.(ssa.Node), maxInstrReached, lastBlockVisited, false)
+
+	// The actual integer Index should not be visited.
+	// Everything but the actual integer Index should be visited.
+	case *ssa.IndexAddr:
+		s.visitReferrers(n, maxInstrReached, lastBlockVisited)
+		s.dfs(t.X.(ssa.Node), maxInstrReached, lastBlockVisited, false)
+
+	// Only the Map itself can be tainted by an Update.
+	// The Key can't be tainted.
+	// The Value can propagate taint to the Map, but not receive it.
+	// MapUpdate has no referrers, it is only an Instruction, not a Value.
```

Comment:
Certainly out of scope for this PR, possibly out of scope for this analyzer, but do we need to consider where keys are tainted?  E.g.

```golang
func TestFoo() {
	s := core.Source{Data: "password1234"}
	m := map[core.Source]string{s: s.Data}
	for src, str := range m {
		core.Sink(src) // want "a source has reached a sink"
		core.Sink(str) // want "a source has reached a sink"
	}
}
```

Reply:
Hmm. With this case in mind, I think in general we may want to consider *anything* coming out of a tainted map to be tainted. The current tests don't do that, e.g.

```go
func TestRangeOverMapWithSourceAsKey() {
	m := map[core.Source]string{core.Source{Data: "password1234"}: "don't sink me"}
	for src, str := range m {
		core.Sink(src) // want "a source has reached a sink"
		core.Sink(str)
	}
}
```

In the above cases, the fix would be to propagate through the `Next` instruction. In the current PR, we do not propagate through those instructions at all; on master, we do. I think for now it would be best to add that case you mentioned and to keep propagating through `Next` instructions.

(For `MapUpdate` though I'm confident we don't want to traverse to anything other than the `Map`).

Reply:
(I'm going to use sources in keys and strings in values in this comment.  The same logic applies in reverse.)

I think if we propagate through `Next`, we might end up tainting keys categorically where we shouldn't.  It could be worth a shot and seeing how it plays out though.

But the more I think about it, the more I think this might be out of scope.  It feels like we're needing to decide categorically whether or not both keys and values are tainted if a map contains a source type value.  Otherwise, we would have to make a distinction between the map "safely" holding its values versus having been tainted by receiving a tainted value as a key.

Reply:
SG. I've added it to #188.